lib/TPP/ConvertPackOptimization.cpp

```diff
@@ -0,0 +1,202 @@
+////===ConvertPackOptimization.cpp ----------------------------*---- C++-*-===//
+////
+//// Part of the LLVM Project, under the Apache License v2.0 with LLVM
+/// Exceptions. / See https://llvm.org/LICENSE.txt for license information. /
+/// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+////
+////===----------------------------------------------------------------------===//
+//
+//
+
+#include "TPP/BuilderUtils.h"
+#include "TPP/Passes.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace mlir;
+using namespace mlir::tpp;
+
+#define GEN_PASS_CLASSES
+#include "TPP/Passes.h.inc"
+
+namespace {
+
+struct MatchResult {
+  bool haveMatch;
+  int index;
+};
+
+/*
+ * Returns a list of indices that match the inner dimensions' position
+ */
+SmallVector<MatchResult> innerDimMatchesIndex(tensor::PackOp packOp) {
+  SmallVector<MatchResult> result;
+  int innerTiles = 0;
+  for (size_t i = 0; i < packOp.getSourceType().getShape().size(); i++) {
+    if (packOp.getInnerDimsPos().size() > 0) {
+      bool haveMatch = false;
+      for (size_t j = 0; j < packOp.getInnerDimsPos().size(); j++) {
+        if (i == (size_t)packOp.getInnerDimsPos()[j]) {
+          MatchResult tempResult;
+          tempResult.haveMatch = true;
+          haveMatch = true;
+          tempResult.index = innerTiles++;
+          result.push_back(tempResult);
+          break;
+        }
+      }
+      if (!haveMatch) {
+        MatchResult tempResult;
+        tempResult.haveMatch = false;
+        tempResult.index = -1;
+        result.push_back(tempResult);
+      }
+    }
+  }
+  return result;
+}
+
+struct ConvertPackOptimizationOp : public OpRewritePattern<tensor::PackOp> {
+  using OpRewritePattern<tensor::PackOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(tensor::PackOp packOp,
+                                PatternRewriter &rewriter) const override {
+    if (packOp.getStaticInnerTiles().size() <= 0)
+      return failure();
+    auto shape = packOp.getSourceType().getShape();
+    int numLoops = shape.size();
+
+    SmallVector<MatchResult> haveMatched = innerDimMatchesIndex(packOp);
+    for (int i = 0; i < numLoops; i++) {
+      if (haveMatched[i].haveMatch) {
+        if (shape[i] < packOp.getStaticInnerTiles()[haveMatched[i].index]) {
```

Comment:
why do you check this?

Reply:
This is in case the returned index is -1.

Reply:
%pack1 = tensor.pack %input_tensor_bcast4 inner_dims_pos = [3, 2] inner_tiles = [8, 32] into %packed_tensor1 : tensor<1x1x32x8xf32> -> tensor<1x1x1x1x8x32xf32>
In this case, inner_tiles size is 32 whereas corresponding shape is 8.

Reply:
Here you are using a tile of 8 for the position in the input tensor at index 3 that is 8, similarly you use a 32 tile factor for the position in the input tensor at index 2 that is 32. The condition above in my opinion is not correct.

Reply:
Replaced with monotonically increasing dimensions check